Src/WindowsServer/WindowsServer.Shared/Implementation/AzureMetadataRequestor.cs

```diff
@@ -0,0 +1,109 @@
+        /// <returns>An instance of AzureInstanceComputeMetadata or null.</returns>
+        private Func<string, Task<AzureInstanceComputeMetadata>> azureIMSRequestor = null;
+
+        internal AzureMetadataRequestor(Func<string, Task<AzureInstanceComputeMetadata>> makeAzureIMSRequestor = null)
+        {
+            this.azureIMSRequestor = makeAzureIMSRequestor;
+        }
+
+        /// <summary>
+        /// Gets or sets the base URI for the Azure Instance Metadata service. Internal to allow overriding in test.
+        /// </summary>
+        internal string BaseAimsUri { get; set; } = "http://169.254.169.254/metadata/instance/compute";
+
+        public Task<AzureInstanceComputeMetadata> GetAzureComputeMetadataAsync()
+        {
+            string metadataRequestUrl = string.Format(
+                CultureInfo.InvariantCulture,
+                "{0}?{1}&{2}",
+                this.BaseAimsUri,
+                AzureMetadataRequestor.AzureImsJsonFormat,
+                AzureMetadataRequestor.AzureImsApiVersion);
+
+            return this.MakeAzureMetadataRequestAsync(metadataRequestUrl);
+        }
+
+        private async Task<AzureInstanceComputeMetadata> MakeAzureMetadataRequestAsync(string metadataRequestUrl)
+        {
+            AzureInstanceComputeMetadata requestResult = null;
+
+            SdkInternalOperationsMonitor.Enter();
+            try
+            {
+                if (this.azureIMSRequestor != null)
+                {
+                    requestResult = await this.azureIMSRequestor(metadataRequestUrl).ConfigureAwait(false);
+                }
+                else
+                {
+                    requestResult = await this.MakeWebRequestAsync(metadataRequestUrl).ConfigureAwait(false);
+                }
+            }
+            catch (Exception ex)
+            {
+                WindowsServerEventSource.Log.AzureInstanceMetadataRequestFailure(
+                    metadataRequestUrl, ex.Message, ex.InnerException != null ? ex.InnerException.Message : string.Empty);
+            }
+            finally
+            {
+                SdkInternalOperationsMonitor.Exit();
+            }
+
+            return requestResult;
+        }
+        
+        private async Task<AzureInstanceComputeMetadata> MakeWebRequestAsync(string requestUrl)
+        {
+            AzureInstanceComputeMetadata azureIms = null;
+            DataContractJsonSerializer deserializer = new DataContractJsonSerializer(typeof(AzureInstanceComputeMetadata));
+
+            using (var azureImsClient = new HttpClient())
+            {
+                azureImsClient.MaxResponseContentBufferSize = AzureMetadataRequestor.AzureImsMaxResponseBufferSize;
+                azureImsClient.DefaultRequestHeaders.Add("Metadata", "True");
+                azureImsClient.Timeout = this.AzureImsRequestTimeout;
+
+                Stream content = await azureImsClient.GetStreamAsync(requestUrl).ConfigureAwait(false);
```

Comment:
Consider:
```C#
using( Stream content = await azureImsClientEtc... )
{
   azureIms = etc...
}
```
(On the assumption that ReadObject or the cast could throw)

Reply:
Yes, I was thinking that way when you originally brought this up. However, I've seen in the past that nesting using blocks can cause problems, [a quick search results in finding articles such as this](https://superdevelopment.com/2014/03/13/common-pitfalls-with-idisposable-and-the-using-statement/).

Reply:
I skimmed through the article; nothing wrong with it, but I don't think it applies here because the two using statements aren't connected. Even so, ReadObject or the cast could throw, so use try/finally.